### DISCUSSION TITLE
Сбой выполнения

### DISCUSSION THREAD

**Discussion by Bondian** [Chimildic/goofy#228] [Помощь с ошибкой]
Чаще чем раз в неделю в логах наблюдаю такой сбой выполнения 15 минутного триггера:

27 нояб. 2022 г., 10:16:35	Ошибка	TypeError: Cannot read property 'id' of null
    at [unknown function]((   0 library:1193:38)
    at removeTracks((   0 library:1192:39)
    at myHistory(1 ( ИСТОРИЯ ) мин 15:8:10)

Не знаю насколько это безобидно...

```
function myHistory() {
  let nowHistory = HistoryManager.getTracks();                      // Беру свежую историю
  let archive_0_History = Cache.read('SpotifyRecentTracks.json');   // Беру архив-0 истории
  let archive_1_History = Cache.read('SpotifyRecentTracks_1.json'); // Беру архив-1 истории
  let dopHistory = Source.getPlaylistTracks('ПРОСЛУШАНО', '29pHiCUfD0wgZCZ4aFKLNB'); // ручное добавление
  Playlist.removeTracks('29pHiCUfD0wgZCZ4aFKLNB', dopHistory);      // Самоочищение "ПРОСЛУШАНО"
  Combiner.push(nowHistory, dopHistory);                            // Объединяю свежую историю и ПРОСЛУШАНО
  Filter.removeTracks(nowHistory, archive_0_History);               // Удаляю из Свежей ист архив-0 истории
  Filter.removeTracks(nowHistory, archive_1_History);               // Удаляю из Свежей ист архив-1 истории
  Filter.dedupTracks(nowHistory);                                   // Удаляю дубликаты из Свежей истории
  Cache.compressTracks(nowHistory);                                 // Сжимаю Свежую историю
  Cache.append('SpotifyRecentTracks.json', nowHistory);             // Добавляю Свежую историю в архив-0
  
  // Cache.append('SpotifyRecentTracks_1.json', nowHistory); // Записывать сюда при наполнении первого файла
  
  console.log('Число запросов', CustomUrlFetchApp.getCountRequest());

}
```

**Comment by Chimildic** [comment]
В массивы треков попадает пустой трек. По логу ошибки не пойму на какой строке. Потому что `at myHistory(1 ( ИСТОРИЯ ) мин 15:8:10)` выглядит странно. Не изменяли? 

Возможно от `HistoryManager.getTracks` приходит раз она самая свежая. В таком случае попробуйте:
```js
let nowHistory = HistoryManager.getTracks().filter(t => !t)
```

**Reply by Bondian** [reply]
1 ( ИСТОРИЯ ) мин 15 - это такое замысловатое имя файла:
(скрин)
![Скриншот 27-11-2022 143339](https://user-images.githubusercontent.com/85184404/204128403-898a7f2e-2ab3-4333-bd39-b8aa027452e1.png)

![Скриншот 27-11-2022 144148](https://user-images.githubusercontent.com/85184404/204128713-70231db6-ef9f-4ccc-bc6b-6613fa96a821.png)


Попробую изменить строку как указано. Понаблюдаю...

**Reply by Chimildic** [reply]
Тогда получается на 8 строке `Filter.removeTracks(nowHistory, archive_0_History)`. Кэш вряд ли _null_ поймал, остается как раз `nowHistory` от `HistoryManager`. Правда туда еще залетает `dopHistory`. Будет странно, если каждую неделю в него попадает сломанный трек, который по API не взять.

**Reply by Bondian** [reply]
Да. dopHistory сейчас не использутся, но и не мешает... Это был ручной костыль, когда я не понимал, почему в своё время HistoryManager не работал.... Потом выяснилось, что обновлённые версии приложения больше не позволяют считывать "быструю" историю... Так и сижу на старой версии.

**Reply by Bondian** [reply]
> В таком случае попробуйте:
> ```js
> let nowHistory = HistoryManager.getTracks().filter(t => !t)
> ```
Не знаю как именно это изменение работает, но теперь из HistoryManager треки не попадают в файл SpotifyRecentTracks.json

**Reply by Chimildic** [reply]
Проверить так хотел. Ошибки про `removeTracks` были? 
```js
// Так попадут
let nowHistory = HistoryManager.getTracks().filter(t => t != undefined)
```

**Reply by Bondian** [reply]
Ошибок не было.
Спасибо, исправил. Недельку буду наблюдать.

**Reply by amd64fox** [reply]
> Так и сижу на старой версии.

Если это еще актуально, то могу добавить новый способ накопления истории, тестирую уже около недели на ласт приложениях android и windows.

**Comment by Bondian** [comment]
> Если это еще актуально, то могу добавить новый способ накопления истории

> 
Конечно актуально. С нетерпением жду )
Ласт не пользую, windows редко, а вот андроид - постоянно...

**Reply by amd64fox** [reply]
[добавил](https://github.com/Chimildic/goofy/discussions/230)
